user-console/src/main/java/org/pentaho/mantle/client/ui/tabs/MantleTabPanel.java

```diff
@@ -58,12 +66,75 @@ public class MantleTabPanel extends org.pentaho.gwt.widgets.client.tabs.PentahoT
   private static final String FRAME_ID_PRE = "frame_"; //$NON-NLS-1$
   private static int frameIdCount = 0;
 
+  private static MenuBar tabsMenuBar;
+  private static MenuItem tabsMenuItem;
+  private static TabContextMenuBar tabsSubMenuBar;
+
+  private static final String EMPTY_TABS_MENU = "empty-tabs-menu";
+
+  private HashMap<PentahoTab, MantleTabMenuItem> menuItemHashMap;
+
+  private static class TabContextMenuBar extends com.google.gwt.user.client.ui.MenuBar {
+
+    String height, maxHeight;
+    int left, top, menuBarTop;
+
+    public TabContextMenuBar( boolean b ) {
+      super( b );
+    }
+
+    @Override
+    public void onBrowserEvent( Event event ) {
+      super.onBrowserEvent( event );
+
+      MenuItem item = MenuBarUtils.findItem( this, DOM.eventGetTarget( event ) );
+      int type = event.getTypeInt();
+      switch ( type ) {
+        case Event.ONMOUSEOVER: {
+          if ( item != null ) {
+            DecoratedPopupPanel popup = MenuBarUtils.getPopup( this );
+            left = this.getOffsetWidth();
+            top = item.getAbsoluteTop();
+            menuBarTop = this.getAbsoluteTop();
+            if ( top < menuBarTop ) {
+              top = menuBarTop + 1;
+            }
+            height = "calc( 100vh - " + top + "px )";
+            maxHeight = MenuBarUtils.calculatePopupHeight( item.getSubMenu() ) + "px";
+            popup.getWidget().getElement().getStyle().setProperty( "height", height );
+            popup.getWidget().getElement().getStyle().setProperty( "maxHeight", maxHeight );
+            popup.setPopupPosition( left, top );
+          }
+          break;
+        }
+      }
+    }
+  }
+
+  private class MantleTabMenuItem extends com.google.gwt.user.client.ui.MenuItem {
```

Comment:
something to think about

Have you thought about moving these small inner clasess` MantleTabMenuItem` `TabContextMenuBar,` etc..  into their own files. MantleTabPanel is approaching 1k lines. I've seen industry standards from 500-800 LOC.

https://eslint.org/docs/latest/rules/max-lines#:~:text=While%20there%20is%20not%20an,from%20100%20to%20500%20lines.